### ISSUE TITLE
Live streaming: Source quality and no automatic source resolution transcoding

### ISSUE THREAD

**Issue by fenarinarsa** [Chocobozzz/PeerTube#5173]
### Describe the problem to be solved

When doing live stream, Peertube only has options to transcode all streams or to transcode none.
The first solution adds heavy loads to the CPU even if the source stream is okay for most viewers. The second one forbids to create low bitrate streams to improve the live streaming experience for clients that have a low bandwidth.

When transcoding is enabled, let say you send a 3Mbps 1080p stream in rtmp and checked "480" in transcoding settings.
Peertube will compress two streams: 480 _and_ 1080, even though the admin didn't request a 1080 compression. It's actually the behavior of the VOD transcoder, but it doesn't fit live streamings.

In addition, transcoding a 1080 stream puts a heavy load on the CPU, which actually makes streaming impossible on a lot of hardware setups. Then the only way is to disable transcoding, which kinds of defeats the whole purpose.

I did some tests on my hardware setup (server in DC) and it can handle 480 compression easily but not 1080+480.

### Describe the solution you would like

The transcoding behavior for live streams should be different than the VOD one.

1. Transcoding into resolutions that are checked **only**
2. Adding an option to include source quality (no recompression). Note that the source compression must be live-compatible, but it usually is when using streaming software.

It means that if "480 + source quality" is checked, peertube will compress 480 **only** and add the source stream in the HLS playlist.

The source resolution compression must not be done unless explicitly checked in the admin setup (for instance, 1080 compression won't happen unless 1080 is checked in the resolutions list). Meaning that in any case, peertube will NOT automatically add the source resolution into the compression loop as it's currently done.

It can be achieved with that kind of ffmpeg call, which is basically the current one + a stream copy. In this case, I took the 1080 resolution out of the filter_complex parameter (I guess it's generated through a loop) and added the "copy" stream.

`ffmpeg -i rtmp://source -y -filter_complex "[v:0]scale=w=-2:h=480[vout480]" -threads 4 -sc_threshold 0 -max_muxing_queue_size 1024 -map_metadata -1 -pix_fmt yuv420p -map "[vout480]" -level:v:0 3.1 -g:v:0 60 -c:v:0 libx264 -preset veryfast -maxrate 1474560 -bufsize 2949120 -b_strategy 1 -bf 16 -r:v:0 30 -b:v:0 1474560 -map a:0 -c:a:0 copy -map v:0 -c:v:1 copy -map a:0 -c:a:1 copy -hls_time 4 -hls_list_size 15 -hls_flags delete_segments+independent_segments+program_date_time -hls_segment_filename %v-%06d.ts -master_pl_name master.m3u8 -f hls -var_stream_map 'v:0,a:0 v:1,a:1' %v.m3u8`

**Comment by Chocobozzz** [comment]
Hello,

I'm currently working on https://github.com/Chocobozzz/PeerTube/issues/1030 and will do the same for live streams.

`transcoding.bound_to_enabled_resolution: false` is added to the configuration. If you enable it PeerTube will only transcode to resolutions you chose. The only exception is if chosen resolutions are above the original resolution (you enable 480p but the video is in 240p so PeerTube will transcode it to 240p).

What do you think about that?

**Comment by fenarinarsa** [comment]
@Chocobozzz  Hello, this does not seem to be fixed with #1030 , which does not address the feature I was talking of: keeping the original stream resolution (no transcoding) and adding another transcoded smaller ones for live streams. So that for smaller servers it only has to transcode a small stream and not the big original one that can be copied as is in the HLS playlist by ffmpeg.

I just installed 4.3 and we still only still have the option to keep the original stream (1 stream) or have every stream transcoded.

Both issues are closed so either I missed something, or you decided not to implement this feature?

**Comment by fenarinarsa** [comment]
My tests never showed any issue with streams generated with OBS (I do even think most services like Twitch actually do a stream copy for the "source" stream). My opinion is that it's up to the user to use a compatible client.

I don't get how to make plugins for peertube since I'm not a good js dev enough to make packages etc. The best I could come with is to make a dirty patch directly peertube's to add the source stream. I actually copy/pasted the looped code that generated the ffmpeg commend line and hardcoded the source stream into that piece of code to generate the ffmpeg command I gave in my first post. It worked fine, the only issue was that the player didn't switch to source quality when it was possible, for no obvious reason. But the player do have some strange behaviors related to bitrate/quality detection even with no stream patch anyway. I reverted back to the regular peertube code after that because I didn't want to leave dirty code in production.

Anyway thanks for your reply, currently peertube is not a streaming solution I'm willing to use in production anymore for many other reasons. Hopefully it'll get better.
